### PR TITLE
HBHE: arrival time

### PR DESCRIPTION
**Pull request by mariadalfonso**
Implementation of a new arrival time definition, more robust when we have PU

Original authors Jayashri Padmanaban and  Jae Hyeok Yoo https://indico.cern.ch/event/1142347/contributions/4793749/attachments/2412936/4129323/HCAL%20timing%20update.pdf 

only activate offline.
complete the issue open here https://github.com/cms-sw/cmssw/issues/28423

@igv4321  @abdoulline

Ideally aiming to have in RECO of Run3 MC 12-4

### CODE REVIEW

**Comment by mariadalfonso** [comment]
enable profiling

**Comment by mariadalfonso** [comment]
please test

**Comment by abdoulline** [comment]
@mariadalfonso   I'd suggest the title 
HCAL: HB/HE signal arrival time  (?)

**Comment by mariadalfonso** [comment]
please test

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc [hunk 1]
```diff
@@ -345,6 +356,78 @@ void MahiFit::updateCov(const SampleMatrix& samplecov) const {
   nnlsWork_.covDecomp.compute(invCovMat);
 }
 
+float MahiFit::ccTime(const float itQ) const {
+  float ccTime_ = 0.;
```
Comment (mandrenguyen) [inline RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc:360]: Why use a trailing subscript for ccTime?

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc [hunk 2]
```diff
@@ -345,6 +356,78 @@ void MahiFit::updateCov(const SampleMatrix& samplecov) const {
...
+
+  // those conditions are now on data time slices, can be done on the fitted pulse i.e. using nlsWork_.ampVec.coeff(itIndex);
+
+  const int soi = nnlsWork_.tsOffset;
+
+  // Selecting energetic hits - (Energy in TS[3] and TS[4]) > 20 GeV
+  const bool cond1 =
```
Comment (mandrenguyen) [inline RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc:367]: If I read correctly, if any of the conditions are false, you would like to return zero.
Why not just do the following?
`if(nnlsWork_.amplitudes.coeffRef(soi) + nnlsWork_.amplitudes.coeffRef(soi + 1U) < thEnergeticPulses_) return 0.;`
and similarly for the other two conditions?
Then you can move the initialization of ccTime to below these 3 conditions.

**Comment by mariadalfonso** [comment]
please test

**Comment by jordan-martins** [comment]
hi @jpata , this is the PR I mentioned in the joint ops meeting. Best, Jordan

**Comment by mandrenguyen** [comment]
The time distribution now seems to have only zero values for a number of distributions, e.g.,
[all_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-08-29-2300+dfcae7/52457/validateJR/all_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999/all_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999c_recoPFClusters_particleFlowClusterHCAL__RECO_obj_time.png)

Is that expected?

**Comment by mandrenguyen** [comment]
please test

**Comment by mandrenguyen** [comment]
> The time distribution now seems to have only zero values for a number of distributions, e.g., [all_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-08-29-2300+dfcae7/52457/validateJR/all_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999/all_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999c_recoPFClusters_particleFlowClusterHCAL__RECO_obj_time.png)
> 
> Is that expected?

My question persists after the last commit, e.g.,
[all_OldVSNew_TTbarPUwf25202p0c_recoPFClusters_particleFlowClusterHCAL__RECO_obj_time](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-08-31-1100+dfcae7/52495/validateJR/all_OldVSNew_TTbarPUwf25202p0/all_OldVSNew_TTbarPUwf25202p0c_recoPFClusters_particleFlowClusterHCAL__RECO_obj_time.png)

**Comment by mariadalfonso** [comment]
The value of the ccTime are as follow

`-9999` when there is no MahiFit possible
`0` when the 3 conditions are not satisfied (let's if we can loose some of these requirements)
~ `- 1BX` as it's now shifter (I will have to correct this)

Here the distribution sin tuba events

**Comment by mandrenguyen** [comment]
@mariadalfonso Ok, so interpret that there will be further commits coming in this PR.  Let me know whether I've misunderstood.

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
please test

(relval input failure)

**Comment by mandrenguyen** [comment]
@mariadalfonso  
The changes to the HCAL time distributions seem to correspond to your description. 
I see this has an effect on the beam-halo filters:
[all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999c_recoBeamHaloSummary_BeamHaloSummary__RECO_obj_HcalLooseHaloId](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-04-2300+dfcae7/52574/validateJR/all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999/all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999c_recoBeamHaloSummary_BeamHaloSummary__RECO_obj_HcalLooseHaloId.png)

Are we ok with that?

**Comment by rappoccio** [comment]
As discussed yesterday during ORP, we will request DQM and DAQ to also review and sign off on this large PR that affects the limited datataking for pp in the remainder of 2022. Once the checks are added, we can merge.

**Comment by mariadalfonso** [comment]
> As discussed yesterday during ORP, we will request DQM and DAQ to also review and sign off on this large PR that affects the limited datataking for pp in the remainder of 2022. Once the checks are added, we can merge.

DAQ ?

This only affect the offline reconstruction ! and the new floats aim to serve LLP analyses

**Comment by mariadalfonso** [comment]
This is the time now in ttbar MC events

**Comment by mariadalfonso** [comment]
> @mariadalfonso The changes to the HCAL time distributions seem to correspond to your description. I see this has an effect on the beam-halo filters: [all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999c_recoBeamHaloSummary_BeamHaloSummary__RECO_obj_HcalLooseHaloId](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-04-2300+dfcae7/52574/validateJR/all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999/all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999c_recoBeamHaloSummary_BeamHaloSummary__RECO_obj_HcalLooseHaloId.png)
> 
> Are we ok with that?

yes, we kind of expect changes like this/



> @mariadalfonso The changes to the HCAL time distributions seem to correspond to your description. I see this has an effect on the beam-halo filters: [all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999c_recoBeamHaloSummary_BeamHaloSummary__RECO_obj_HcalLooseHaloId](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-04-2300+dfcae7/52574/validateJR/all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999/all_mini_OldVSNew_TTbar14TeV2026D88PUwPRMXwf39634p999c_recoBeamHaloSummary_BeamHaloSummary__RECO_obj_HcalLooseHaloId.png)
> 
> Are we ok with that?

We do expect changes here and there, we are moving to a more meaningful definition for high energy hits (before was more something a quantity proportional to the residual of the fits), and for low energy the time is not well measured and should not be used downstream.

**Comment by mandrenguyen** [comment]
please test

**Comment by mariadalfonso** [comment]
please test

(this PR doesn't usa CUDA Flags)

**Comment by mariadalfonso** [comment]
@mandrenguyen can you restart the test ? Thanks

**Comment by mandrenguyen** [comment]
please test

**Comment by mandrenguyen** [comment]
please test

**Comment by mandrenguyen** [comment]
please test
trying again in case the missing input is transient

**Comment by mandrenguyen** [comment]
@cms-sw/jetmet-pog-l2 I would like your opinion on this PR.
Browsing the code repository I get the impression the beam halo filters are used for MET. 
This PR changes the hcal arrival time estimate, which has a downstream effect on the beam halo filters, e.g.,
[HcalLooseHaloId](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-09-2300+dfcae7/52692/validateJR/all_OldVSNew_RunHI2018wf140p56/all_OldVSNew_RunHI2018wf140p56c_recoBeamHaloSummary_BeamHaloSummary__reRECO_obj_HcalLooseHaloId.png) 

Do you see any problem with changing the definition of these filters?  Note that the PR is planned to be backported in 12_4_X for production this year.

**Comment by lathomas** [comment]
Indeed the beam halo filters use the calo time information. I think the impact of the change should be assessed by the MET scanners. @dvannerom
One could for example check fake rate/noise rejection on [/MuonEG/Run2022C-TopMuEG-PromptReco-v1/RAW-RECO](https://cmsweb.cern.ch/das/request?input=dataset%3D%2FMuonEG%2FRun2022C-TopMuEG-PromptReco-v1%2FRAW-RECO&instance=prod/global) and [/MET/Run2022C-HighMET-PromptReco-v1/RAW-RECO](https://cmsweb.cern.ch/das/request?input=dataset%3D%2FMET%2FRun2022C-HighMET-PromptReco-v1%2FRAW-RECO&instance=prod/global)

**Comment by mandrenguyen** [comment]
> Indeed the beam halo filters use the calo time information. I think the impact of the change should be assessed by the MET scanners. @dvannerom One could for example check fake rate/noise rejection on [/MuonEG/Run2022C-TopMuEG-PromptReco-v1/RAW-RECO](https://cmsweb.cern.ch/das/request?input=dataset%3D%2FMuonEG%2FRun2022C-TopMuEG-PromptReco-v1%2FRAW-RECO&instance=prod/global) and [/MET/Run2022C-HighMET-PromptReco-v1/RAW-RECO](https://cmsweb.cern.ch/das/request?input=dataset%3D%2FMET%2FRun2022C-HighMET-PromptReco-v1%2FRAW-RECO&instance=prod/global)

Thanks @lathomas  @dvannerom Let me add that we will discuss this development at this week's reco meeting, which might be a useful forum for further exchange on this topic. 
 https://indico.cern.ch/event/1199969/

**Comment by mandrenguyen** [comment]
urgent
As the 12_4_X backport is requested for resumption of data taking.
It's me holding this up, but once I get go-ahead from MET experts, I will approve. 
Of course the decision to include or not this in 12_4 will presumably need to be discussed in some forum.

**Comment by alkaloge** [comment]
Although the MET scanners is already notified about this, it is not possible to give feedback on such short notice. We will ask it to give high priority, but again, it will take some time.

**Comment by perrotta** [comment]
@mariadalfonso @mandrenguyen it seems to me that this is mostly intended for MC, for which we still have a few days.
But if it is also intended for the data taking, please notice that we plan to cut the release (12_4_X) tomorrow, and if we want to include in it, the master release has to get merged not later than this evening.
Please let us know which are the plans!

**Comment by fwyzard** [comment]
What are the plan for propagating these changes to the GPU implementation of MAHI ?

**Comment by mandrenguyen** [comment]
+reconstruction
Changes are in line with expectation of improved  timing estimate for HCAL. 
Note that we are still waiting for feedback from JetMET on the impact of the change to beam halo filters on MET.

**Comment by perrotta** [comment]
+1
-  let have this merged in master, for now

**Comment by perrotta** [comment]
> What are the plans for propagating these changes to the GPU implementation of MAHI ?

@mariadalfonso @fwyzard I imagine this was discussed at the [reco meeting on Sept 16](https://indico.cern.ch/event/1199969), to which I was not able to connect. Could anyone summarize here the answer, for the record?

**Pull request merged**